### ISSUE TITLE
Opaque template errors when trying to Serialize Thirdparty type 

### ISSUE THREAD

**Issue by kernalphage** [Donerkebap13/DonerSerializer#5]
Hi there, I was trying to serialize `glm::vec4` using the method described in your readme, and ran in to a large tower of template errors. Have you seen something like this? 

Code example is [here](https://gist.github.com/kernalphage/f5ccdd61ddd03d8c3c4c7318d3ffaf89) (I can provide a CMakeList as well, but I'm building on G++ 7.3 and std C++14). Example is done with a locally defined `mytype`, but a similar error occurs when trying to define something more complex. 

Errors start out like this, then stop when it reaches the max template depth of 900:
```
In file included from ~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:30:0,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
/usr/include/c++/7/tuple: In instantiation of ‘class std::tuple_element<18446744073709551615, std::tuple<> >’:
/usr/include/c++/7/utility:133:69:   required by substitution of ‘template<long unsigned int __i, class _Tp> using __tuple_element_t = typename std::tuple_element::type [with long unsigned int __i = 18446744073709551615; _Tp = std::tuple<>]’
/usr/include/c++/7/tuple:1320:5:   required by substitution of ‘template<long unsigned int __i, class ... _Elements> constexpr std::__tuple_element_t<__i, std::tuple<_Elements ...> >& std::get(const std::tuple<_Elements ...>&) [with long unsigned int __i = 18446744073709551615; _Elements = {}]’
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:109:48:   required from ‘void DonerReflection::DoApplyFunctionToObjectElements(T&, Args&& ...) [with long unsigned int iteration = 18446744073709551615; Resolver = DonerSerializer::CSerializationResolver; T = mytype; Args = {rapidjson::GenericDocument<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>, rapidjson::CrtAllocator>&}]’
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:122:55:   required from ‘typename std::enable_if<(iteration > 0)>::type DonerReflection::ApplyFunctionToObjectElements(T&, Args&& ...) [with long unsigned int iteration = 18446744073709551615; Resolver = DonerSerializer::CSerializationResolver; T = mytype; Args = {rapidjson::GenericDocument<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>, rapidjson::CrtAllocator>&}; typename std::enable_if<(iteration > 0)>::type = void]’
~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:217:5:   required from ‘bool DonerSerializer::CJsonSerializer::Serialize(T&) [with T = mytype]’
~/donertest/Definitions.hpp:65:4:   required from ‘static bool Util::load_json(T&, const char*, bool, bool) [with T = mytype]’
~/donertest/main.cpp:6:48:   required from here
/usr/include/c++/7/tuple:1297:7: error: static assertion failed: tuple index is in range
       static_assert(__i < tuple_size<tuple<>>::value,
       ^~~~~~~~~~~~~
In file included from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32:0,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h: In instantiation of ‘void DonerReflection::DoApplyFunctionToObjectElements(T&, Args&& ...) [with long unsigned int iteration = 18446744073709551615; Resolver = DonerSerializer::CSerializationResolver; T = mytype; Args = {rapidjson::GenericDocument<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>, rapidjson::CrtAllocator>&}]’:
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:122:55:   required from ‘typename std::enable_if<(iteration > 0)>::type DonerReflection::ApplyFunctionToObjectElements(T&, Args&& ...) [with long unsigned int iteration = 18446744073709551615; Resolver = DonerSerializer::CSerializationResolver; T = mytype; Args = {rapidjson::GenericDocument<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>, rapidjson::CrtAllocator>&}; typename std::enable_if<(iteration > 0)>::type = void]’
~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:217:5:   required from ‘bool DonerSerializer::CJsonSerializer::Serialize(T&) [with T = mytype]’
~/donertest/Definitions.hpp:65:4:   required from ‘static bool Util::load_json(T&, const char*, bool, bool) [with T = mytype]’
~/donertest/main.cpp:6:48:   required from here
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:109:48: error: no matching function for call to ‘get<18446744073709551615>(const std::tuple<>&)’
   constexpr auto property = std::get<iteration>(std::decay<SDonerReflectionClassProperties<T>>::type::s_properties);
                             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/tuple:38:0,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:30,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
/usr/include/c++/7/utility:225:5: note: candidate: template<long unsigned int _Int, class _Tp1, class _Tp2> constexpr typename std::tuple_element<_Int, std::pair<_Tp1, _Tp2> >::type& std::get(std::pair<_Tp1, _Tp2>&)
     get(std::pair<_Tp1, _Tp2>& __in) noexcept
     ^~~
/usr/include/c++/7/utility:225:5: note:   template argument deduction/substitution failed:
In file included from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32:0,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:109:48: note:   types ‘std::pair<_Tp1, _Tp2>’ and ‘const std::tuple<>’ have incompatible cv-qualifiers
   constexpr auto property = std::get<iteration>(std::decay<SDonerReflectionClassProperties<T>>::type::s_properties);
                             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/tuple:38:0,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:30,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
/usr/include/c++/7/utility:230:5: note: candidate: template<long unsigned int _Int, class _Tp1, class _Tp2> constexpr typename std::tuple_element<_Int, std::pair<_Tp1, _Tp2> >::type&& std::get(std::pair<_Tp1, _Tp2>&&)
     get(std::pair<_Tp1, _Tp2>&& __in) noexcept
     ^~~
/usr/include/c++/7/utility:230:5: note:   template argument deduction/substitution failed:
In file included from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32:0,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:109:48: note:   types ‘std::pair<_Tp1, _Tp2>’ and ‘const std::tuple<>’ have incompatible cv-qualifiers
   constexpr auto property = std::get<iteration>(std::decay<SDonerReflectionClassProperties<T>>::type::s_properties);
                             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/tuple:38:0,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:30,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
/usr/include/c++/7/utility:235:5: note: candidate: template<long unsigned int _Int, class _Tp1, class _Tp2> constexpr const typename std::tuple_element<_Int, std::pair<_Tp1, _Tp2> >::type& std::get(const std::pair<_Tp1, _Tp2>&)
     get(const std::pair<_Tp1, _Tp2>& __in) noexcept
     ^~~
/usr/include/c++/7/utility:235:5: note:   template argument deduction/substitution failed:
In file included from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32:0,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:109:48: note:   ‘const std::tuple<>’ is not derived from ‘const std::pair<_Tp1, _Tp2>’
   constexpr auto property = std::get<iteration>(std::decay<SDonerReflectionClassProperties<T>>::type::s_properties);
                             ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/tuple:38:0,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/thirdparty/DonerReflection/DonerReflection/include/donerreflection/DonerReflection.h:30,
                 from ~/donertest/../ref/DonerSerializer/DonerSerializer/include/donerserializer/DonerSerialize.h:32,
                 from ~/donertest/Definitions.hpp:9,
                 from ~/donertest/main.cpp:1:
/usr/include/c++/7/utility:244:5: note: candidate: template<class _Tp, 
```

**Comment by Donerkebap13** [comment]
Greetings kernalphage!

I'm really happy to see you decided to use DonerSerializer in your project!

About your issue:

You're trying to serialize/deserialize a single variable directly using ``DonerSerializer::CJsonSerializer`` and ``DonerSerializer::CJsonDeserializer``, and **this is not possible**.

This two systems are meant to be used with classes/structs that defines their members info using ``DONER_DEFINE_REFLECTION_DATA`` example [here](https://github.com/Donerkebap13/DonerSerializer#how-to-use-it).

What you're defining in ``CSerializationResolver::CSerializationResolverType<glm::vec4>`` and ``CDeserializationResolver::CDeserializationResolverType<glm::vec4>`` is the way of serializing/deserializing that specific type using the system mentioned above, so for example this would work:

```c++
struct MyUniforms
{
	glm::vec4 m_uniform1;
	glm::vec4 m_uniform2;
}

DONER_DEFINE_REFLECTION_DATA(MyUniforms,
	 DONER_ADD_NAMED_VAR_INFO(m_uniform1, "uniform1"),
	 DONER_ADD_NAMED_VAR_INFO(m_uniform2, "uniform2")
)
```

``DONER_DEFINE_REFLECTION_DATA`` creates a class with an std::tuple with all the members defined in each ``DONER_ADD_NAMED_VAR_INFO`` statement. This tuple is used later on inside ``DonerSerializer::CJsonSerializer`` and ``DonerSerializer::CJsonDeserializer``.

Right now, if you only want to serialize/deserialize your ``glm::vec4``, you need to code a little bit more, using directly the methods you've just created.

For serializing:
```c++
rapidjson::Document document;
document.SetObject();
// This line serializes your 'glm::vec4' inside document
DonerSerializer::CSerializationResolver::CSerializationResolverType<glm::vec4>::Apply("custom_name", obj, document);

// This is the terrible way of getting a string with the generated json using rapidjson
rapidjson::StringBuffer strbuf;
strbuf.Clear();
rapidjson::Writer<rapidjson::StringBuffer> writer(strbuf);
document.Accept(writer);
std::string result = std::string(strbuf.GetString());
// result == { "custom_name":[ 0.0, 1.0, 2.0, 4.0 ] }
```

For deserializing:
```c++
rapidjson::Document parser;
rapidjson::Value& root = parser.Parse(jsonContents.c_str());

// This line will dump the json inside obj
DonerSerializer::CDeserializationResolver::CDeserializationResolverType<glm::vec4>::Apply(obj, root);
```
So to recap: depending on what you need to serialize, consider creating a container struct/class with all in it and define it's data using ``DONER_DEFINE_REFLECTION_DATA``. If you just want to serialize one single ``glm::vec4`` at a time, for the moment use the code written above.

Anyway, I'll think about adding the option of serializing single variables using ``DonerSerializer::CJsonSerializer`` and ``DonerSerializer::CJsonDeserializer`` directly, but I can't give you an estimated time on when that could be in.

Hope I solved your issue!

Cheers and thank you again for using DonerSerializer!

**Comment by kernalphage** [comment]
Yes, wrapping it in a container type worked, thanks! 

Stop by [here](https://github.com/kernalphage/gl_play) if you want to see what I've been up to. Using JSON for saving off my editor state, mostly.
